### PR TITLE
Use keywords for the rest of the files

### PR DESCRIPTION
**Pull request by cakester**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

using keywords for items in config files

### CODE REVIEW

File: ivadomed/training.py [hunk 1]
```diff
@@ -54,7 +55,7 @@ def train(model_params, dataset_train, dataset_val, training_params, path_output
     writer = SummaryWriter(log_dir=path_output)
 
     # BALANCE SAMPLES AND PYTORCH LOADER
-    conditions = all([training_params["balance_samples"]["applied"], model_params["name"] != "HeMIS"])
+    conditions = all([training_params["balance_samples"]["applied"], model_params[ModelParamsKW.NAME] != "HeMIS"])
```
Comment (dyt811) [inline ivadomed/training.py:58]: I think `training_params["balance_samples"]["applied"]` are from 
```
@dataclass
class BalanceSamplesKW:
    APPLIED = "applied"
```

File: ivadomed/inference.py [hunk 2]
```diff
@@ -110,10 +111,11 @@ def get_onehotencoder(context: dict, folder_model: str, options: dict, ds: Datas
     metadata_dict = joblib.load(Path(folder_model, 'metadata_dict.joblib'))
     for idx in ds.indexes:
         for i in range(len(idx)):
-            idx[i]['input_metadata'][0][context['FiLMedUnet']['metadata']] = options['metadata']
+            idx[i]['input_metadata'][0][context[ConfigKW.FILMED_UNET][ModelParamsKW.METADATA]] = options['metadata']
```
Comment (dyt811) [inline ivadomed/inference.py:114]: So I think you missed the keyword conversion for the `options` dict. Which is generated initially at the `run_segment_command` [here](https://github.com/ivadomed/ivadomed/blob/0c2e6d7d0dd16feadd97929b342e1796a6c02b7d/ivadomed/main.py#L232)

File: ivadomed/inference.py [hunk 3]
```diff
@@ -110,10 +111,11 @@ def get_onehotencoder(context: dict, folder_model: str, options: dict, ds: Datas
...
             idx[i]['input_metadata'][0]['metadata_dict'] = metadata_dict
 
-    ds = imed_film.normalize_metadata(ds, None, context["debugging"], context['FiLMedUnet']['metadata'])
+    ds = imed_film.normalize_metadata(
+        ds, None, context[ConfigKW.DEBUGGING], context[ConfigKW.FILMED_UNET][ModelParamsKW.METADATA])
```
Comment (dyt811) [inline ivadomed/inference.py:118]: Might want to eventually add a check for `ConfigKW.FILMED_UNET` field first before invoking its keys.

File: ivadomed/inference.py [hunk 4]
```diff
@@ -240,19 +242,19 @@ def process_transformations(context: dict, fname_roi: str, fname_prior: str, met
...
         logger.warning(
             "fname_roi has not been specified, then a cropping around the center of the image is "
             "performed instead of a cropping around a Region of Interest.")
 
-        context["transformation"] = dict((key, value) if key != 'ROICrop'
-                                         else ('CenterCrop', value)
-                                         for (key, value) in context["transformation"].items())
+        context[ConfigKW.TRANSFORMATION] = dict((key, value) if key != TransformationKW.ROICROP
+                                         else (TransformationKW.CENTERCROP, value)
+                                         for (key, value) in context[ConfigKW.TRANSFORMATION].items())
```
Comment (dyt811) [inline ivadomed/inference.py:252]: This is rather ... hard to read. I recommend breaking it into loops and properly expand it out.

File: ivadomed/testing.py [hunk 5]
```diff
@@ -125,7 +126,7 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
             # input_samples: list of batch_size tensors, whose size is n_channels X height X width X depth
             # gt_samples: idem with n_labels
             # batch['*_metadata']: list of batch_size lists, whose size is n_channels or n_labels
-            if model_params["name"] == "HeMISUnet":
+            if model_params[ModelParamsKW.NAME] == ConfigKW.HEMIS_UNET:
                 input_samples = imed_utils.cuda(imed_utils.unstack_tensors(batch["input"]), cuda_available)
             else:
                 input_samples = imed_utils.cuda(batch["input"], cuda_available)
```
Comment (dyt811) [inline ivadomed/testing.py:132]: `batch` here is a Torch DataLoader  and these are keywords specific for Torch DataLoader so we might want a PR to gather them as well to be consistent.

File: ivadomed/testing.py [hunk 6]
```diff
@@ -180,15 +181,15 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
             if "bounding_box" in batch['input_metadata'][smp_idx][0]:
```
Comment (dyt811) [inline ivadomed/testing.py:181]: Same Batch key

File: ivadomed/testing.py [hunk 7]
```diff
@@ -180,15 +181,15 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
...
                 imed_obj_detect.adjust_undo_transforms(testing_params["undo_transforms"].transforms, batch, smp_idx)
 
-            if model_params["is_2d"]:
+            if model_params[ModelParamsKW.IS_2D]:
                 preds_idx_arr = None
                 idx_slice = batch['input_metadata'][smp_idx][0]['slice_index']
                 n_slices = batch['input_metadata'][smp_idx][0]['data_shape'][-1]
```
Comment (dyt811) [inline ivadomed/testing.py:187]: Just for the `input_metadata` for now.

File: ivadomed/training.py [hunk 8]
```diff
@@ -54,8 +55,8 @@ def train(model_params, dataset_train, dataset_val, training_params, path_output
     writer = SummaryWriter(log_dir=path_output)
 
     # BALANCE SAMPLES AND PYTORCH LOADER
-    conditions = all([training_params["balance_samples"]["applied"], model_params["name"] != "HeMIS"])
-    sampler_train, shuffle_train = get_sampler(dataset_train, conditions, training_params['balance_samples']['type'])
+    conditions = all([training_params["balance_samples"][BalanceSamplesKW.APPLIED], model_params[ModelParamsKW.NAME] != "HeMIS"])
+    sampler_train, shuffle_train = get_sampler(dataset_train, conditions, training_params['balance_samples'][BalanceSamplesKW.TYPE])
```
Comment (dyt811) [inline ivadomed/training.py:59]: You missed a few keywords here. "balanced_samples" is just from the context JSON. 

Also, I this training_params is just from the context dict, as this is the only place it was called: 
image

File: ivadomed/training.py [hunk 9]
```diff
@@ -64,7 +65,7 @@ def train(model_params, dataset_train, dataset_val, training_params, path_output
 
     gif_dict = {"image_path": [], "slice_id": [], "gif": []}
     if dataset_val:
-        sampler_val, shuffle_val = get_sampler(dataset_val, conditions, training_params['balance_samples']['type'])
+        sampler_val, shuffle_val = get_sampler(dataset_val, conditions, training_params['balance_samples'][BalanceSamplesKW.TYPE])
```
Comment (dyt811) [inline ivadomed/training.py:68]: Same here.

File: ivadomed/main.py [hunk 10]
```diff
@@ -73,11 +75,11 @@ def get_parser():
...
-    path_model = Path(path_output, context.get("model_name"))
+    path_model = Path(path_output, context[ConfigKW.MODEL_NAME])
     if not path_model.is_dir():
         logger.info(f'Creating model directory: {path_model}')
         path_model.mkdir(parents=True)
-        if 'film_layers' in model_params and any(model_params['film_layers']):
+        if ModelParamsKW.FILM_LAYERS in model_params and any(model_params[ModelParamsKW.FILM_LAYERS]):
             joblib.dump(train_onehotencoder, path_model.joinpath("one_hot_encoder.joblib"))
             if 'metadata_dict' in ds_train[0]['input_metadata'][0]:
                 metadata_dict = ds_train[0]['input_metadata'][0]['metadata_dict']
```
Comment (dyt811) [inline ivadomed/main.py:85]: We might want to start thinking about pulling these BidsDataset keywords out eventually as well in a future PR as well. This one is already getting incredibly large.

File: ivadomed/main.py [hunk 11]
```diff
@@ -204,29 +212,30 @@ def update_film_model_params(context, ds_test, model_params, path_output):
...
 
 def run_segment_command(context, model_params):
     # BIDSDataframe of all image files
     # Indexing of derivatives is False for command segment
-    bids_df = BidsDataframe(context['loader_parameters'], context['path_output'], derivatives=False)
+    bids_df = BidsDataframe(
+        context[ConfigKW.LOADER_PARAMETERS], context[ConfigKW.PATH_OUTPUT], derivatives=False)
 
     # Append subjects filenames into a list
     bids_subjects = sorted(bids_df.df.get('filename').to_list())
```
Comment (dyt811) [inline ivadomed/main.py:232]: Do we have a Keyword class for `bids_df`'s class dataframe column name

File: ivadomed/main.py [hunk 12]
```diff
@@ -235,15 +244,15 @@ def run_segment_command(context, model_params):
     seen_subj_ids = []
 
     for subject in bids_subjects:
-        if context['loader_parameters']['multichannel']:
+        if context[ConfigKW.LOADER_PARAMETERS][LoaderParamsKW.MULTICHANNEL]:
             # Get subject_id for multichannel
             df_sub = bids_df.df.loc[bids_df.df['filename'] == subject]
```
Comment (dyt811) [inline ivadomed/main.py:249]: Add this to the ticket above.

File: ivadomed/main.py [hunk 13]
```diff
@@ -263,8 +272,9 @@ def run_segment_command(context, model_params):
             fname_img = bids_df.df[bids_df.df['filename'] == subject]['path'].to_list()
 
         # Add film metadata to options for segment_volume
-        if 'film_layers' in model_params and any(model_params['film_layers']) and model_params['metadata']:
-            metadata = bids_df.df[bids_df.df['filename'] == subject][model_params['metadata']].values[0]
+        if ModelParamsKW.FILM_LAYERS in model_params and any(model_params[ModelParamsKW.FILM_LAYERS]) \
+                and model_params[ModelParamsKW.METADATA]:
+            metadata = bids_df.df[bids_df.df['filename'] == subject][model_params[ModelParamsKW.METADATA]].values[0]
             options['metadata'] = metadata
```
Comment (dyt811) [inline ivadomed/main.py:278]: Here `options` are the same one I mentioned about dictionary used during run segmentation commands

File: ivadomed/main.py [hunk 14]
```diff
@@ -352,26 +363,28 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
     bids_df = BidsDataframe(loader_params, path_output, derivatives=True)
 
     # Get subject filenames lists. "segment" command uses all participants of data path, hence no need to split
-    train_lst, valid_lst, test_lst = imed_loader_utils.get_subdatasets_subject_files_list(context["split_dataset"],
+    train_lst, valid_lst, test_lst = imed_loader_utils.get_subdatasets_subject_files_list(context[ConfigKW.SPLIT_DATASET],
                                                                                           bids_df.df,
                                                                                           path_output,
-                                                                                          context["loader_parameters"]
+                                                                                          context[ConfigKW.LOADER_PARAMETERS]
                                                                                           ['subject_selection'])
```
Comment (dyt811) [inline ivadomed/main.py:370]: Missed one here.

File: ivadomed/main.py [hunk 15]
```diff
@@ -430,7 +443,7 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
                                'training')
 
         # Choice of optimisation metric
-        metric = "recall_specificity" if model_params["name"] in imed_utils.CLASSIFIER_LIST else "dice"
+        metric = "recall_specificity" if model_params[ModelParamsKW.NAME] in imed_utils.CLASSIFIER_LIST else "dice"
```
Comment (dyt811) [inline ivadomed/main.py:446]: Could you also extract these two keywords? as the metrics keywords: `recall_specificity` and `dice`. There might be more.

File: ivadomed/main.py [hunk 16]
```diff
@@ -475,17 +488,17 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
...
-                                              target_suffix=loader_params["target_suffix"],
-                                              eval_params=context["evaluation_parameters"])
+                                              target_suffix=loader_params[LoaderParamsKW.TARGET_SUFFIX],
+                                              eval_params=context[ConfigKW.EVALUATION_PARAMETERS])
         return df_results, pred_metrics
 
 
 def create_dataset_and_ivadomed_version_log(context):
-    path_data = context.get('loader_parameters').get('path_data')
+    path_data = context[ConfigKW.LOADER_PARAMETERS][LoaderParamsKW.PATH_DATA]
```
Comment (dyt811) [inline ivadomed/main.py:501]: Again, not sure why took out the old `.get()` approach.

**Review by dyt811 — CHANGES_REQUESTED** [review]
Please carefully review the comments as there are probably additional PRs needed to capture the other changes in place. 

There were quite a few things to adjust namely missed keywords. 

Once comments are addressed, ping me for the second round of review.

File: ivadomed/inference.py [hunk 17]
```diff
@@ -75,8 +76,8 @@ def get_preds(context: dict, fname_model: str, model_params: dict, gpu_id: int,
             model.eval()
 
             # Films/Hemis based prediction require meta data load
-            if ('FiLMedUnet' in context and context['FiLMedUnet']['applied']) or \
-                    ('HeMISUnet' in context and context['HeMISUnet']['applied']):
+            if (ConfigKW.FILMED_UNET in context and context[ConfigKW.FILMED_UNET][ModelParamsKW.APPLIED]) or \
+                    (ConfigKW.HEMIS_UNET in context and context[ConfigKW.HEMIS_UNET][ModelParamsKW.APPLIED]):
```
[the remark this answers is not in the document]
Comment (konantian) [reply ivadomed/inference.py]: Done.

File: ivadomed/inference.py (hunk 3 above)
Comment (konantian) [reply ivadomed/inference.py]: Done

File: ivadomed/training.py (hunk 8 above)
Comment (konantian) [reply ivadomed/training.py]: Done

File: ivadomed/training.py [hunk 18]
```diff
@@ -54,8 +55,8 @@ def train(model_params, dataset_train, dataset_val, training_params, path_output
     writer = SummaryWriter(log_dir=path_output)
 
     # BALANCE SAMPLES AND PYTORCH LOADER
-    conditions = all([training_params["balance_samples"]["applied"], model_params["name"] != "HeMIS"])
-    sampler_train, shuffle_train = get_sampler(dataset_train, conditions, training_params['balance_samples']['type'])
+    conditions = all([training_params["balance_samples"][BalanceSamplesKW.APPLIED], model_params[ModelParamsKW.NAME] != "HeMIS"])
+    sampler_train, shuffle_train = get_sampler(dataset_train, conditions, training_params['balance_samples'][BalanceSamplesKW.TYPE])
 
     train_loader = DataLoader(dataset_train, batch_size=training_params["batch_size"],
```
[the remark this answers is not in the document]
Comment (konantian) [reply ivadomed/training.py:63]: Done

File: ivadomed/testing.py (hunk 5 above)
Comment (dyt811) [reply ivadomed/testing.py:132]: Currently only taking care of minor batch keywords.

**Review by dyt811 — APPROVED** [review]
